File: Core kernel Yaiwes/OpenAgentd/code/app/api/schemas/mcp.py
"""Request and response schemas for ``/api/mcp`` endpoints."""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.agent.mcp.config import HttpServerConfig, OAuthConfig, StdioServerConfig


class StdioServerBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    transport: Literal["stdio"] = "stdio"
    command: Annotated[str, Field(min_length=1)]
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)
    enabled: bool = True

    def to_config(self) -> StdioServerConfig:
        return StdioServerConfig(
            command=self.command,
            args=self.args,
            env=self.env,
            enabled=self.enabled,
        )


class OAuthBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    client_id: str | None = None
    client_secret: str | None = None

    def to_config(self) -> OAuthConfig:
        return OAuthConfig(
            client_id=self.client_id,
            client_secret=self.client_secret,
        )


class HttpServerBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    transport: Literal["http"] = "http"
    url: Annotated[str, Field(min_length=1)]
    headers: dict[str, str] = Field(default_factory=dict)
    oauth: OAuthBody | None = None
    enabled: bool = True

    def to_config(self) -> HttpServerConfig:
        return HttpServerConfig(
            url=self.url,
            headers=self.headers,
            oauth=self.oauth.to_config() if self.oauth else None,
            enabled=self.enabled,
        )


ServerBody = StdioServerBody | HttpServerBody


class CreateServerRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Annotated[str, Field(min_length=1)]
    server: ServerBody = Field(discriminator="transport")


class UpdateServerRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    server: ServerBody = Field(discriminator="transport")


class ServerStatusResponse(BaseModel):
    """Live runner state plus the saved config from ``mcp.json``."""

    name: str
    transport: str
    enabled: bool
    state: str
    error: str | None = None
    tool_names: list[str] = Field(default_factory=list)
    started_at: str | None = None
    config: ServerBody | None = Field(default=None, discriminator="transport")


class ServerListResponse(BaseModel):
    servers: list[ServerStatusResponse]


class ServerDeleteResponse(BaseModel):
    name: str


class MCPAppToolCallRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    session_id: str
    tool_call_id: str
    server: Annotated[str, Field(min_length=1)]
    tool: Annotated[str, Field(min_length=1)]
    arguments: dict[str, Any] = Field(default_factory=dict)


class MCPAppToolCallResponse(BaseModel):
    result: dict[str, Any]

File: Core kernel Yaiwes/OpenAgentd/code/web/src/api/client/mcp.ts
/**
 * OpenAgentd API client — MCP group: /mcp servers + app tools.
 */

import { apiBaseUrl } from '../base-url'
import { parseDetailOrThrow } from './_shared'

export type StdioServerBody = {
  transport: 'stdio'
  command: string
  args: string[]
  env: Record<string, string>
  enabled: boolean
}

export type HttpServerBody = {
  transport: 'http'
  url: string
  headers: Record<string, string>
  oauth?: {
    client_id?: string | null
    client_secret?: string | null
  } | null
  enabled: boolean
}

export type ServerBody = StdioServerBody | HttpServerBody

export type ServerStatus = {
  name: string
  transport: 'stdio' | 'http'
  enabled: boolean
  state: 'stopped' | 'starting' | 'ready' | 'auth_required' | 'error'
  error: string | null
  tool_names: string[]
  started_at: string | null
  /** Saved config from mcp.json. Null when the server was removed mid-flight. */
  config: ServerBody | null
}

export type CreateServerRequest = { name: string; server: ServerBody }
export type UpdateServerRequest = { server: ServerBody }
export type ServerDeleteResponse = { name: string }
export type MCPAppToolCallRequest = {
  session_id: string
  tool_call_id: string
  server: string
  tool: string
  arguments: Record<string, unknown>
}
export type MCPAppToolCallResponse = { result: unknown }

export async function listMcpServers(): Promise<{ servers: ServerStatus[] }> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers`)
  if (!res.ok) await parseDetailOrThrow(res, 'listMcpServers')
  return res.json()
}

export async function getMcpServer(name: string): Promise<ServerStatus> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers/${encodeURIComponent(name)}`)
  if (!res.ok) await parseDetailOrThrow(res, `GET /mcp/servers/${name}`)
  return res.json()
}

export async function createMcpServer(name: string, server: ServerBody): Promise<ServerStatus> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name, server }),
  })
  if (!res.ok) await parseDetailOrThrow(res, 'POST /mcp/servers')
  return res.json()
}

export async function updateMcpServer(name: string, server: ServerBody): Promise<ServerStatus> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers/${encodeURIComponent(name)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ server }),
  })
  if (!res.ok) await parseDetailOrThrow(res, `PUT /mcp/servers/${name}`)
  return res.json()
}

export async function deleteMcpServer(name: string): Promise<ServerDeleteResponse> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers/${encodeURIComponent(name)}`, { method: 'DELETE' })
  if (!res.ok) await parseDetailOrThrow(res, `DELETE /mcp/servers/${name}`)
  return res.json()
}

export async function restartMcpServer(name: string): Promise<ServerStatus> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers/${encodeURIComponent(name)}/restart`, { method: 'POST' })
  if (!res.ok) await parseDetailOrThrow(res, `POST /mcp/servers/${name}/restart`)
  return res.json()
}

export async function connectMcpOAuth(name: string): Promise<ServerStatus> {
  const res = await fetch(`${apiBaseUrl()}/mcp/servers/${encodeURIComponent(name)}/oauth/connect`, { method: 'POST' })
  if (!res.ok) await parseDetailOrThrow(res, `POST /mcp/servers/${name}/oauth/connect`)
  return res.json()
}

export async function callMcpAppTool(body: MCPAppToolCallRequest): Promise<MCPAppToolCallResponse> {
  const res = await fetch(`${apiBaseUrl()}/mcp/app-tools/call`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) await parseDetailOrThrow(res, 'POST /mcp/app-tools/call')
  return res.json()
}
